### PR TITLE
ci(ruff): Enforce the default `C901` complexity

### PR DESCRIPTION
**Pull request by dangotbanned**
# References
- https://docs.astral.sh/ruff/rules/complex-structure/#why-is-this-bad
- https://docs.astral.sh/ruff/settings/#lint_mccabe_max-complexity

Previously was set at **18**, the default is **10**.
As the rule was not in `[tool.ruff.lint.select]`, the setting was never used.
Setting such a high limit hides opportunities to simplify large, complex functions.

Many cases were resolved/improved in vega#3431.
There were 2 I fixed in this PR as they were quite simple changes.
The remainder use the *# noqa: C901* comment to serve as a TODO.

### CODE REVIEW

File: altair/jupyter/jupyter_chart.py [hunk 1]
```diff
@@ -218,7 +218,7 @@ def __init__(
         )
 
     @traitlets.observe("chart")
-    def _on_change_chart(self, change):
+    def _on_change_chart(self, change):  # noqa: C901
```
Comment (dangotbanned) [inline altair/jupyter/jupyter_chart.py:221]: I think this was the highest, at around 20

File: altair/vegalite/v5/api.py [hunk 2]
```diff
@@ -3876,7 +3876,8 @@ def _check_if_valid_subspec(
             raise ValueError(err.format(attr, classname))
 
 
-def _check_if_can_be_layered(spec: LayerType | dict[str, Any]) -> None:
+# C901 fixed in https://github.com/vega/altair/pull/3520
+def _check_if_can_be_layered(spec: LayerType | dict[str, Any]) -> None:  # noqa: C901
```
Comment (dangotbanned) [inline altair/vegalite/v5/api.py:3880]: See the resolution for this in #3520

**Comment by jonmmease** [comment]
Thanks for working through these rules @dangotbanned!

I'll defer to others, but I don't personally find C901 all that helpful. We should of course use functions and classes to minimize duplication of logic, but for an inherently complex function that doesn't contain logic that's useful elsewhere, I don't personally get much benefit from it being broken down into smaller functions.  And I suppose this is reflected in my coding style 😄

**Comment by dangotbanned** [comment]
> Thanks for working through these rules @dangotbanned!
> 
> I'll defer to others, but I don't personally find C901 all that helpful. We should of course use functions and classes to minimize duplication of logic, but for an inherently complex function that doesn't contain logic that's useful elsewhere, I don't personally get much benefit from it being broken down into smaller functions. And I suppose this is reflected in my coding style 😄

Totally valid point @jonmmease! Appreciate your thoughts.

Choosing to use `C901` (or any rule really) is ultimately opt-in per *"violation"*.
I shouldn't have phrased it in the description as *every* case needing to be *"fixed"*.

Personally, the benefit to me when writing new code that the warning prompts me to consider *if* there is an alternative way to write the function.

> `function` is too complex (15 > 10)

If there isn't, or if it is time sensitive - add a *# noqa: C901* and move on.
The important part is the **moment of pause** the rule introduces, which *for me* usually leads to thinking more critically about my own code

**Comment by jonmmease** [comment]
Yeah, it's easy to disable when needed. So no objection, and the small changes you needed to make to pass the check all look positive to me.

**Review by jonmmease — APPROVED** [review]
Looks good. Feel free to merge when you're ready!

**Comment by dangotbanned** [comment]
> Looks good. Feel free to merge when you're ready!

Thanks @jonmmease!

FYI if at any time you find it frustrating to add a lot of ignore comments (for a single file), you can always add to our [tool.ruff.lint.per-file-ignores](https://github.com/dangotbanned/altair/blob/13c06a7ff6585cad278af540ca6b512544d7ae17/pyproject.toml#L388-L392) table instead

https://docs.astral.sh/ruff/settings/#lint_per-file-ignores

**Pull request merged**